### PR TITLE
Feature Calculator

### PR DESCRIPTION
**Pull request by ridera7**
Created Calculator

### CODE REVIEW

File: src/main/java/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+public class Calculator {
```
Comment (yura-jt) [inline src/main/java/Calculator.java:1]: Please always use package.
You can not reference to class without package.
If you add some package to your application, and then try to use Calculator, you will fail.

com.calculator

com.calculator.app
com.calculator.modul 

etc.

Package should be present always. It's a must.

In your case

com.calculator enough

Usually it's just reprsent application name:

com.watch.shop.application
com.medecine.handler
com.fitness.booking.service
com.booking.app
сom.calculator

File: src/main/java/Calculator.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    public double multiply(double a1, double a2){
+        return a1 * a2;
+    }
+
+    public double divide(double a1, double a2) throws IllegalArgumentException{
+        if (a2 == 0) throw new IllegalArgumentException("The divisor cannot be equal to 0");
+        return a1 / a2;
+    }
+
+}
```
Comment (yura-jt) [inline src/main/java/Calculator.java:20]: POM.xml:

<groupId>com.calculator</groupId>
not Ridera/Task/example etc, or company name if it's commercial product, or if it's your personal project, just package structure

I put blank line betwen <properties> and <dependencies> so we could seprate it and increase readability

Also, in pom.xml and in java class as well, before </..> closing tag or before closing brace } we DO  NOT put empty line.

There is exception in java class, we CAN put empty line before CLASS closing brace:
`
    public double divide(double a1, double a2) throws IllegalArgumentException{
        if (a2 == 0) throw new IllegalArgumentException("The divisor cannot be equal to 0");
        return a1 / a2;
    }

}`
In this case it's ok


`<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.calculator</groupId>
    <artifactId>calculator</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>21</maven.compiler.source>
        <maven.compiler.target>21</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.11.0</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>`

File: src/main/java/Calculator.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+public class Calculator {
+
+    public double add(double a1, double a2){
```
Comment (yura-jt) [inline src/main/java/Calculator.java:3]: a1, a2 is acceptable naming, but has it's drawbacks:
It's more complex because we use TWO alphanumeric systems:
a, b, c, d..
and
1, 2...

We can reduce complexity just by using mathematic notation for arithmetic operations:
a and b:
```
  public double subtract(double a, double b){
        return a - b;
    }
```
It's a little bit more easier to read

File: src/main/java/Calculator.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        return a1 + a2;
+    }
+
+    public double subtract(double a1, double a2){
```
Comment (yura-jt) [inline src/main/java/Calculator.java:7]: double is ok, since no requirements were for it.

But you can make it a little bit simpler, just use int every where.
Only in divide:
`public double divide(int a, int b)`

File: src/main/java/Calculator.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        return a1 - a2;
+    }
+
+    public double multiply(double a1, double a2){
```
Comment (yura-jt) [inline src/main/java/Calculator.java:11]: Please time from time use CTRL+ALT+L reformat code and CTRL + ALT + O optimize imports shortcut.
Especially before commiting or show to someone your code.

Code should be formatted, Idea could do it for you in one shortcut.

`(double a1, double a2){  -> (double a1, double a2) { `

File: src/main/java/Calculator.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        return a1 * a2;
+    }
+
+    public double divide(double a1, double a2) throws IllegalArgumentException{
+        if (a2 == 0) throw new IllegalArgumentException("The divisor cannot be equal to 0");
```
Comment (yura-jt) [inline src/main/java/Calculator.java:16]: IF ELSE ELSE IF, FOR, WHILE etc always should be with braces.
So:
```
 if (b == 0) {
   throw new IllegalArgumentException("The divisor cannot be equal to 0");
 }
```
Two main reasons:
1. Readibility.
 Just compare:
 if (calcAvgMarginByX(45, arg2) || hasTrueBetween(89, 90)) calcAvgMarginByY(axis);
 
 and 
```
 if (calcAvgMarginByX(45, arg2) || hasTrueBetween(89, 90)) {
    calcAvgMarginByY(axis);
}   
```
 In first case we can't just clearly see when condition ends and when expression starts.
 
 2. Second reason it's harder to refactor, If I want to add additional logic , add one more line, I need first just to put braces and only then add my line:
 1st step:
  if (b == 0) {
   throw new IllegalArgumentException("The divisor cannot be equal to 0");
 }
 2nd step:
  if (b == 0) {
   addMyLine();
   throw new IllegalArgumentException("The divisor cannot be equal to 0");
 }
 
 A lot of bugs could be occured there, because it's not clearly where braces { } should be

File: src/test/java/CalculatorTest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (yura-jt) [inline src/test/java/CalculatorTest.java:5]: import static org.junit.jupiter.api.Assertions.*;
тут копіпастну свій комент з іншого ревю
 
Wildcard imports are not allowed in majority of cases,
please read
Wildcards pro and cons - https://www.baeldung.com/java-wildcard-imports
 
There are + and - for each approach, usually projects don't like .* imports with two main reasons:
conflicts with same name classes in different packages (in big projects regulary occurs)
loosing self-descriptiveness.
 
Java code is tending to be self-descriptive, and when we open class and imports we could see EXACTLY what is used in class, it's often helps,
if we see .* we don't know maybe 1000 imports used, maybe 1, what exactly, it ends up with loosing readibility.
 
Also I would recommend to change Idea settings - [intellij-idea-how-to-disable-wildcard-imports](https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports)

File: src/test/java/CalculatorTest.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+class CalculatorTest {
+
+    private Calculator calculator;
+
+    @BeforeEach
+    void setUp() {
+        calculator = new Calculator();
+    }
+
+    @AfterEach
```
Comment (yura-jt) [inline src/test/java/CalculatorTest.java:16]: This method is not needed in our case, just remove it

File: src/test/java/CalculatorTest.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Test
+    void multiply() {
+        assertEquals(26,calculator.multiply(13,2));
+    }
+
+    @Test
+    void divide() {
+        assertEquals(2, calculator.divide(6, 3));
```
Comment (yura-jt) [inline src/test/java/CalculatorTest.java:37]: 1 unit test -  1 call to testing method.

1 unit test can have multiple assert:
RiderEntity actual = service.someMethod(arg);

assertNotNull(actual);
assertEquals("Rider", actual.getName);
assertTrue(actual.isValid());

etc.

So multiple asserts could just examine different sides of SINGLE output / result.

But 1 unit test cover 1 scenario.

Your test cover 2 scenarios - postive and negative, and therefore have 2 method calls.

So there should be two unit tests:

@Test
void shoudlCorrectlyDivide() {
 assertEquals(5, calculator.divide(25, 5));
}

@Test
void shouldThrowExceptionWhenDivideByZeroIsPerformed() {
   IllegalArgumentException ex = assertThrows(IllegalArgumentException.class, () -> {calculator.divide(3,0)});
   
   assertEquals("message...", ex.getMessage());
 }
 
 Don't forget to verify exception message.

File: src/test/java/CalculatorTest.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        assertEquals(7, calculator.add(3, 4));
+    }
+
+    @Test
+    void subtract() {
+        assertEquals(3, calculator.subtract(7,4));
+    }
+
+    @Test
+    void multiply() {
```
Comment (yura-jt) [inline src/test/java/CalculatorTest.java:31]: unit test name should correspond what exactly you testing

A lot of naming convention are present, just use one

checkIfCalculatorDivideCorrectlyLargeNumbers
 
given_validNumbers_shouldCorrectlyDivide
 
shouldThrowExceptionWhenDivideByZero
shouldCorrectlyDivideLargeNumbers

Few samples.

I prefer last one shoud...
but you could use any convenient for you.

File: src/test/java/CalculatorTest.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        calculator = new Calculator();
+    }
+
+    @AfterEach
+    void tearDown() {
+    }
+
+    @Test
+    void add() {
+        assertEquals(7, calculator.add(3, 4));
```
Comment (yura-jt) [inline src/test/java/CalculatorTest.java:22]: 1 test is not enough in this case.

We have simple case 3, 4
We have special number - zero, so 10, 0
Also we have large numbers 25004095 5089470
also we have negative numbers 560900 -546

That's a sample approach, when you write unit test, 
cover positive scenario
cover all condition if there are any
cover negative scenario (exception)

And if you could see different subset of arguments, also check corner / edge case.

Also, to save your time, you could not write 4 unit test for 4 operations , but just use @ParameterizedTest with @CsvSource or @MethodSource
please refer to https://www.baeldung.com/parameterized-tests-junit-5

So CalculatorTest class will have only 4 Parameterized tests for positive scenarios (add, substract etc) + 1 test for exception divide by zero

File: src/main/java/Calculator.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
+public class Calculator {
+
```
Comment (yura-jt) [inline src/main/java/Calculator.java:2]: as for branch name feature-calculator

Usually for feautre most common format:
feature/....

so 
feature/calculator
feature/add-calc-functionality
feature/ABC-563  (if there are JIRA number)
feature/add-book-service

You could do nothing in this case, just get know for future
